### PR TITLE
[BEAM-1347] Break apart ProcessBundleHandler to use service locator pattern based upon URNs.

### PR DESCRIPTION
**Pull request by lukecwik**
This cleans up ProcessBundleHandler and allows for separate improvements of the various PTransform handler factories.

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @vikkyrk

File: sdks/java/harness/src/main/java/org/apache/beam/runners/core/PTransformRunnerFactory.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      Consumer<ThrowingRunnable> addStartFunction,
+      Consumer<ThrowingRunnable> addFinishFunction) throws IOException;
+
+  /**
+   * A registrar which can return a mapping from {@link RunnerApi.FunctionSpec#getUrn()} to
+   * a factory capable of instantiating an appropriate handler.
+   */
+  interface Registrar {
+    /**
+     * Returns a mappring from {@link RunnerApi.FunctionSpec#getUrn()} to a a factory capable of
```
Comment (vikkyrk) [inline sdks/java/harness/src/main/java/org/apache/beam/runners/core/PTransformRunnerFactory.java]: s/mappring/mapping. s/a a/a

File: sdks/java/harness/src/test/java/org/apache/beam/runners/core/BeamFnDataReadRunnerTest.java [hunk 2]
```diff
@@ -190,6 +262,19 @@ public void run() {
...
+    verifyNoMoreInteractions(mockBeamFnDataClient);
+  }
+
+  @Test
+  public void testRegistration() {
+    for (Registrar registrar :
+        ServiceLoader.load(Registrar.class)) {
+      if (registrar instanceof BeamFnDataReadRunner.Registrar) {
+        assertThat(registrar.getPTransformRunnerFactories(),
+            IsMapContaining.hasKey("urn:org.apache.beam:source:runner:0.1"));
```
Comment (vikkyrk) [inline sdks/java/harness/src/test/java/org/apache/beam/runners/core/BeamFnDataReadRunnerTest.java]: Make this a constant in the Handler class. Same at other places.

Comment (lukecwik) [inline sdks/java/harness/src/test/java/org/apache/beam/runners/core/BeamFnDataReadRunnerTest.java]: Done

File: sdks/java/harness/src/main/java/org/apache/beam/runners/core/PTransformRunnerFactory.java (hunk 1 above)
Comment (lukecwik) [inline sdks/java/harness/src/main/java/org/apache/beam/runners/core/PTransformRunnerFactory.java]: Done

**Pull request closed without merging**